lib/saluki-io/src/deser/framing/iter.rs

```diff
@@ -0,0 +1,334 @@
+use std::mem::ManuallyDrop;
+
+use ouroboros::self_referencing;
+
+use super::{Framer, FramingError, RawBuffer};
+use crate::buf::ReadIoBuffer;
+
+struct DirectIter<'framer, 'buf> {
+    framer: &'framer dyn Framer,
+    buf: &'buf [u8],
+    is_eof: bool,
+}
+
+impl<'framer, 'buf> DirectIter<'framer, 'buf> {
+    fn new(framer: &'framer dyn Framer, buf: &'buf [u8], is_eof: bool) -> Self {
+        Self { framer, buf, is_eof }
+    }
+
+    fn buf_len(&self) -> usize {
+        self.buf.len()
+    }
+
+    fn next_frame(&mut self) -> Result<Option<&[u8]>, FramingError> {
+        // Our buffer is empty, so we're all done.
+        if self.buf.is_empty() {
+            return Ok(None);
+        }
+
+        match self.framer.next_frame(RawBuffer::new(self.buf), self.is_eof)? {
+            Some(frame) => {
+                // Advance our buffer past the frame we just extracted.
+                self.buf = &self.buf[frame.buf_len()..];
+                Ok(Some(frame.into_bytes()))
+            }
+            None => Ok(None),
+        }
+    }
+}
+
+struct NestedIter<'framer, 'buf> {
+    outer_framer: &'framer dyn Framer,
+    inner_framer: &'framer dyn Framer,
+    buf: &'buf [u8],
+    outer_frame_buf: &'buf [u8],
+    is_eof: bool,
+}
+
+impl<'framer, 'buf> NestedIter<'framer, 'buf> {
+    fn new(
+        outer_framer: &'framer dyn Framer, inner_framer: &'framer dyn Framer, buf: &'buf [u8], is_eof: bool,
+    ) -> Self {
+        Self {
+            outer_framer,
+            inner_framer,
+            buf,
+            outer_frame_buf: &[],
+            is_eof,
+        }
+    }
+
+    fn buf_len(&self) -> usize {
+        self.buf.len()
+    }
+
+    fn next_frame(&mut self) -> Result<Option<&[u8]>, FramingError> {
+        // Check if our outer frame is empty, and try to extract the next outer frame if so.
+        if self.outer_frame_buf.is_empty() {
+            // If our root buffer is also empty, then we're done.
+            if self.buf.is_empty() {
+                return Ok(None);
+            }
+
+            match self.outer_framer.next_frame(RawBuffer::new(self.buf), self.is_eof)? {
+                Some(frame) => {
+                    // Advance our root buffer past the outer frame we just extracted.
+                    self.buf = &self.buf[frame.buf_len()..];
+
+                    self.outer_frame_buf = frame.into_bytes();
+                }
+                None => return Ok(None),
+            }
+        }
+
+        // NOTE: This should never happen, based on what we're doing above.. but it gives us an invariant that we depend on
+        // below, where if we get no inner frame, then we know the outer frame is corrupted somehow (or we have a framer bug).
+        assert!(!self.outer_frame_buf.is_empty(), "outer frame buf should not be empty");
```

Comment:
Not sure if we actually need to handle this, but it's possible we could get a zero-length frame and blow up here on a "technically correct" input.